### ISSUE TITLE
prefer-template rule parameter for "never"

### ISSUE THREAD

**Issue by shellscape** [eslint/eslint#9301]
**Tell us about your environment**

* **ESLint Version:** 4.4.1
* **Node Version:** 8.2.1
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
We're using this and extending it https://github.com/webpack-contrib/eslint-config-webpack

<details>
<summary>Configuration</summary>

```js
{
  "extends": "webpack"
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
const foo = 'bar';
const res = foo + ' something';
```

`eslint-config-webpak` contains `"prefer-template": "error"` and `"no-useless-concat": "error"`. Nothing out of the ordinary there. In the project I'm working on, we have a few files which are served to a client (browser). When we ran the --fix option, ESLint went ahead and replaced all of the concats in our client files with template literals.

**What did you expect to happen?**

Well we expected that behavior.

**What actually happened? Please include the actual, raw output from ESLint.**

What we didn't realize was that IE (not Edge) wouldn't support the change as it doesn't support template literals. The onus was on us to know the rules that ESLint would apply. However, now that we've identified the problem, we need to go back through and fix the template literals, changing them back to regular concats. Unfortunately, `prefer-template` doesn't have a rule parameter for reversing that automatic change.

Something along the lines of `"prefer-template": ["error", "never"]` would be immensely useful for reverting those changes by way of an `.eslintrc` in the client files directory, and rerunning with --fix.

**Comment by platinumazure** [comment]
I'm not 100% opposed, but I'll point out that this is a solid use case for source control.

That said, the rule does seem incomplete without this option. I'll support this. Let's see what the rest of the team thinks.

**Comment by not-an-aardvark** [comment]
Would this option disallow template literals entirely, or just disallow template literals that contain placeholders? It seems like it would be more consistent for the rule to just disallow literals with placeholders, but it seems like if you want to disallow all template literals, it wouldn't be sufficient. To me, this seems like a case where the [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax) rule would be a better fit (or you could set `ecmaVersion: 5` to disallow template literals in the parser).

**Comment by shellscape** [comment]
@not-an-aardvark preferably it'd disallow template literals altogether. out specific use case is IE11, where template literals aren't supported, but const, let, and some other bits are. `ecmaVersion: 5` would restrict the code to a point that we wouldn't be able to use any 6 features. It's certainly an option should the team deem this request unneeded, but we'd really enjoy picking off and preventing the features we can't use, versus locking down IE11 to ES5.

For completeness, I believe this rule should have the `"never"` parameter, but I'm happy to use `no-restricted-syntax` if someone can point me at the full documentation for the syntax identifiers, as I can't seem to track that down.

**Comment by not-an-aardvark** [comment]
The spec for the AST nodes is [here](https://github.com/estree/estree). For convenience, it might be easiest to use https://astexplorer.net/ so that you can see the node types of your own code. (In this case, I think you're looking for the `TemplateLiteral` node.)

**Comment by shellscape** [comment]
I just looked into the `quotes` rule and it does cover backtick as a quoting type. So that rule takes care of the template literal without placeholders. I believe given that, the proposed parameter would reasonably only target template literals with placeholders.

**Comment by not-an-aardvark** [comment]
That makes sense, although the `quotes` rule doesn't catch template strings with linebreaks in them, because those can't be converted directly to single or double quotes:

```js
var foo = `
  bar
  baz
`;
```

Overall, I'm not necessarily opposed to this proposal, but I don't think it's the best solution to your problem. `prefer-template` is designed around enforcing a particular style of string concatenation, not forbidding particular types of syntax entirely, so it feels like this proposal lacks a motivating use case.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.

**Comment by shellscape** [comment]
@not-an-aardvark I'd be curious to know if the entire team was directed to put eyes on this before it was closed.

**Comment by not-an-aardvark** [comment]
Team members generally look at open issues in the issue tracker when they have time. We get a lot of issues and everyone does this in their spare time, so it wouldn't be feasible to have every team member look at every issue.

**Comment by j-f1** [comment]
@shellscape You might be interested in [`eslint-plugin-compat`](https://github.com/amilajack/eslint-plugin-compat), which will warn you when you use features not supported in the targets you choose.

**Comment by shellscape** [comment]
@j-f1 much better response than the previous one. much appreciated, that'll get me where I need to be.
